service/history/processing/interface.go

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//go:generate mockgen -copyright_file ../../../LICENSE -package $GOPACKAGE -source $GOFILE -destination interface_mock.go -self_package github.com/uber/cadence/service/history/processing
+
+package processing
+
+import (
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/persistence"
+	"github.com/uber/cadence/service/history/task"
+)
+
+type (
+	// DomainFilter filters domain
+	DomainFilter struct {
+		Domains map[string]struct{}
+		// by default, a DomainFilter matches domains listed in the Domains field
+		// if reverseMatch is true then the DomainFilter matches domains that are
+		// not in the Domains field.
+		ReverseMatch bool
+	}
+
+	// JobInfo indicates the scope of a Job and its progress
+	JobInfo interface {
+		QueueID() int
+		MinLevel() task.Key
+		MaxLevel() task.Key
+		AckLevel() task.Key
+		ReadLevel() task.Key
+		DomainFilter() DomainFilter
+	}
+
+	// Job is responsible for keeping track of the state of tasks
+	// within the scope defined by its JobInfo; it can also be split
+	// into multiple Jobs with non-overlapping scope or be merged with
+	// another Job
+	Job interface {
+		Info() JobInfo
+		Split(JobSplitPolicy) []Job
+		Merge(Job) []Job
+		AddTasks(map[task.Key]task.Task)
+		UpdateAckLevel()
+		// TODO: add Offload() method
+	}
+
+	// JobSplitPolicy determins if a Job should be split
+	// into multiple Jobs
+	JobSplitPolicy interface {
+		Evaluate(Job) []JobInfo
+	}
+
+	// JobQueue manages a list of non-overlapping Jobs
+	// and keep track of the current active Job
+	JobQueue interface {
+		Info() []JobInfo
```

Comment:
Why does this return a list of JobInfos instead of a list of Jobs? 

Also, I'd consider renaming this function. It should probably be the plural form of the types of objects returned (ie. Jobs for [|Job, JobInfos for []JobInfo, etc.)

Reply:
I think returning `JobInfo` covers all the use cases I can think of:
- JobManager wants to persist job metadata to DB. In this case, all it needs can be retrieved from JobInfo
- For debugging purpose, we want to "describe the queue". Again in this case all we need is the metadata.

Maybe you have another use case in your mind that requires other information for Job and is not covered by JobInfo?

In terms of renaming, sure I can create type alias like Jobs for []Job, JobInfos for []JobInfo.

Reply:
In my view, JobInfo is essentially a data model object that represents the state/scope of a Job with getters/setters to access and manipulate the fields contained in it. There is a 1:1 relationship between a Job and JobInfo. Due to the nature of the relationship between them, it makes sense to use composition pattern to embed a JobInfo in the Job. I think Job should be the entity that's responsible for updating/persisting its JobInfo when needed, otherwise, bugs can arise due to lack of coordination between the state of Job and JobInfo objects. 

JobQueue is the abstraction that keeps track of a set of Jobs, and each Job object has its JobInfo as discussed above. What I find a bit unusual about the current proposal is that a JobQueue implementation would either need to keep track of JobInfo objects in addition to the Job objects, or it'd need to aggregate them from the Job objects at the time of Info() []JobInfo function call. I think it should simply return those Job objects so that the caller of this function can get the JobInfo from the Job for read-only purposes or interact with the Job interface to modify it (ie. UpdateAckLevel)

Reply:
Synced offline. Idea is make Job itself ignorant of persistence layer, but will change the method signature to `Queues() []ProcessingQueue` to be more flexible.